// provide-js/dist/cjs/clients/rpc-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RpcClient = void 0;
const api_client_1 = require("./api-client");
class RpcClient {
    constructor(scheme = RpcClient.DEFAULT_SCHEME, host = RpcClient.DEFAULT_HOST, path = RpcClient.DEFAULT_PATH, version = RpcClient.DEFAULT_VERSION) {
        this.apiClient = new api_client_1.ApiClient(undefined, scheme, host, path);
        this.id = 1;
        this.version = version;
    }
    call(method, params) {
        return new Promise((resolve, reject) => {
            this.apiClient.post('', {
                id: this.id++,
                version: this.version,
                method: method,
                params: params,
            }).then((resp) => {
                if (resp && resp.data && resp.status === 200) {
                    if (resp.data && typeof resp.data.result !== 'undefined' && typeof resp.data.error === 'undefined') {
                        resolve(resp.data.result);
                    }
                    else if (resp.data && resp.data.error) {
                        reject(resp);
                    }
                    else {
                        reject(resp);
                    }
                }
                else {
                    reject(resp);
                }
            }).catch((err) => {
                reject(err);
            });
        });
    }
}
exports.RpcClient = RpcClient;
RpcClient.DEFAULT_SCHEME = 'http';
RpcClient.DEFAULT_HOST = 'localhost:8545';
RpcClient.DEFAULT_PATH = '/';
RpcClient.DEFAULT_VERSION = '2.0';
