// provide-js/dist/cjs/services/ident.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.identClientFactory = exports.Ident = void 0;
const clients_1 = require("../clients");
class Ident {
    constructor(token, scheme, host, path) {
        if (!host) {
            host = Ident.DEFAULT_HOST;
        }
        this.client = new clients_1.ApiClient(token, scheme, host, path);
    }
    static clientFactory(token, scheme, host, path) {
        const _scheme = scheme ? scheme : (process.env['IDENT_API_SCHEME'] || 'https');
        const _host = host ? host : (process.env['IDENT_API_HOST'] || Ident.DEFAULT_HOST);
        const _path = path ? path : (process.env['IDENT_API_PATH'] || clients_1.ApiClient.DEFAULT_PATH);
        return new Ident(token, _scheme, _host, _path);
    }
    static unauthenticatedClientFactory(token, scheme, host, path) {
        const _scheme = scheme ? scheme : (process.env['IDENT_API_SCHEME'] || 'https');
        const _host = host ? host : (process.env['IDENT_API_HOST'] || Ident.DEFAULT_HOST);
        const _path = path ? path : (process.env['IDENT_API_PATH'] || clients_1.ApiClient.DEFAULT_PATH);
        return new clients_1.ApiClient(token, _scheme, _host, _path);
    }
    static authenticate(params, scheme, host, path) {
        return Ident.unauthenticatedClientFactory(undefined, scheme, host, path).post('authenticate', params);
    }
    static createUser(params, scheme, host, path) {
        return Ident.unauthenticatedClientFactory(undefined, scheme, host, path).post('users', params);
    }
    static requestPasswordReset(email, scheme, host, path) {
        return Ident.unauthenticatedClientFactory(undefined, scheme, host, path).post('reset_password', { email });
    }
    static resetPassword(token, password, scheme, host, path) {
        return Ident.unauthenticatedClientFactory(undefined, scheme, host, path).post(`reset_password/${token}`, { password });
    }
    createApplication(params) {
        return this.client.post('applications', params);
    }
    updateApplication(appId, params) {
        return this.client.put(`applications/${appId}`, params);
    }
    fetchApplications(params) {
        return this.client.get('applications', params);
    }
    fetchApplicationDetails(appId) {
        return this.client.get(`applications/${appId}`, {});
    }
    fetchApplicationOrganizations(appId, params) {
        return this.client.get(`applications/${appId}/organizations`, params);
    }
    createApplicationOrganization(appId, params) {
        return this.client.post(`applications/${appId}/organizations`, params);
    }
    updateApplicationOrganization(appId, organizationId, params) {
        return this.client.put(`applications/${appId}/organizations/${organizationId}`, params);
    }
    deleteApplicationOrganization(appId, organizationId) {
        return this.client.delete(`applications/${appId}/organizations/${organizationId}`);
    }
    fetchApplicationInvitations(appId, params) {
        return this.client.get(`applications/${appId}/invitations`, params);
    }
    fetchApplicationTokens(appId) {
        return this.client.get(`applications/${appId}/tokens`, {});
    }
    fetchApplicationUsers(appId, params) {
        return this.client.get(`applications/${appId}/users`, params);
    }
    createApplicationUser(appId, params) {
        return this.client.post(`applications/${appId}/users`, params);
    }
    updateApplicationUser(appId, userId, params) {
        return this.client.put(`applications/${appId}/users/${userId}`, params);
    }
    deleteApplicationUser(appId, userId) {
        return this.client.delete(`applications/${appId}/users/${userId}`);
    }
    createOrganization(params) {
        return this.client.post('organizations', params);
    }
    fetchOrganizations(params) {
        return this.client.get('organizations', params);
    }
    fetchOrganizationDetails(organizationId) {
        return this.client.get(`organizations/${organizationId}`, {});
    }
    updateOrganization(organizationId, params) {
        return this.client.put(`organizations/${organizationId}`, params);
    }
    fetchOrganizationInvitations(organizationId, params) {
        return this.client.get(`organizations/${organizationId}/invitations`, params);
    }
    fetchOrganizationUsers(organizationId, params) {
        return this.client.get(`organizations/${organizationId}/users`, params);
    }
    createOrganizationUser(organizationId, params) {
        return this.client.post(`organizations/${organizationId}/users`, params);
    }
    updateOrganizationUser(organizationId, userId, params) {
        return this.client.put(`organizations/${organizationId}/users/${userId}`, params);
    }
    deleteOrganizationUser(organizationId, userId) {
        return this.client.delete(`organizations/${organizationId}/users/${userId}`);
    }
    fetchOrganizationVaults(organizationId, params) {
        return this.client.get(`organizations/${organizationId}/vaults`, params);
    }
    fetchOrganizationVaultKeys(organizationId, vaultId, params) {
        return this.client.get(`organizations/${organizationId}/vaults/${vaultId}/keys`, params);
    }
    createOrganizationVaultKey(organizationId, vaultId, params) {
        return this.client.post(`organizations/${organizationId}/vaults/${vaultId}/keys`, params);
    }
    deleteOrganizationVaultKey(organizationId, vaultId, keyId) {
        return this.client.delete(`organizations/${organizationId}/vaults/${vaultId}/keys/${keyId}`);
    }
    organizationVaultKeySignMessage(organizationId, vaultId, keyId, msg) {
        return this.client.post(`organizations/${organizationId}/vaults/${vaultId}/keys/${keyId}/sign`, { message: msg });
    }
    organizationVaultKeyVerifySignature(organizationId, vaultId, keyId, msg, sig) {
        return this.client.post(`organizations/${organizationId}/vaults/${vaultId}/keys/${keyId}/verify`, { message: msg, signature: sig });
    }
    fetchOrganizationVaultSecrets(organizationId, vaultId, params) {
        return this.client.get(`organizations/${organizationId}/vaults/${vaultId}/secrets`, params);
    }
    createOrganizationVaultSecret(organizationId, vaultId, params) {
        return this.client.post(`organizations/${organizationId}/vaults/${vaultId}/secrets`, params);
    }
    deleteOrganizationVaultSecret(organizationId, vaultId, secretId) {
        return this.client.delete(`organizations/${organizationId}/vaults/${vaultId}/secrets/${secretId}`);
    }
    createToken(params) {
        return this.client.post('tokens', params);
    }
    fetchTokens(params) {
        return this.client.get('tokens', params);
    }
    fetchTokenDetails(tokenId) {
        return this.client.get(`tokens/${tokenId}`, {});
    }
    deleteToken(tokenId) {
        return this.client.delete(`tokens/${tokenId}`);
    }
    createInvitation(params) {
        return this.client.post('invitations', params);
    }
    createUser(params) {
        return this.client.post('users', params);
    }
    fetchUsers() {
        return this.client.get('users', {});
    }
    fetchUserDetails(userId) {
        return this.client.get(`users/${userId}`, {});
    }
    updateUser(userId, params) {
        return this.client.put(`users/${userId}`, params);
    }
}
exports.Ident = Ident;
Ident.DEFAULT_HOST = 'ident.provide.services';
exports.identClientFactory = (token, scheme, host, path) => {
    return Ident.clientFactory(token, scheme, host, path);
};
